### PR TITLE
Automatically refresh when registration opens

### PR DESCRIPTION
**Pull request by FinnIckler**
https://github.com/user-attachments/assets/f59f58d5-a716-4bec-bbd5-49e8d5e8bf78

Uses the old message if the competition opens in more than one hour, then it switches to the countdown.

### CODE REVIEW

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import React, { useEffect, useState } from 'react';
+import { Message } from 'semantic-ui-react';
+import I18nHTMLTranslate from '../../I18nHTMLTranslate';
+
+export default function RegistrationClosedMessage({
+  registrationStart,
+  onTimerEnd,
+}) {
+  const [timeLeft, setTimeLeft] = useState({
+    days: 0, hours: 0, minutes: 0, seconds: 0,
```
Comment (kr-matthews) [inline app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:11]: This should have the correct initial value. Especially since there's a `timeLeft.hours < 1` check below.

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      const diff = start.diff(now, ['days', 'hours', 'minutes', 'seconds']).toObject();
+      setTimeLeft({
+        days: Math.floor(diff.days),
+        hours: Math.floor(diff.hours),
+        minutes: Math.floor(diff.minutes),
+        seconds: Math.floor(diff.seconds),
+      });
+    };
+
+    const intervalId = setInterval(updateTimer, 1000); // Update every second
```
Comment (kr-matthews) [inline app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:35]: I'm not sure how important this is for performance or anything, but if there's more than 1 hour left then you could set the next update to happen when there's an hour left. Although then the `days`, `hours`, `minutes`, and `seconds` values will be incorrect, so maybe this is a bad suggestion.

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    updateTimer(); // Initial call to set the state right away
+
+    return () => clearInterval(intervalId); // Cleanup interval on unmount
+  }, [registrationStart, onTimerEnd, start]);
+
+  return (
+    <Message color="blue">
+      { timeLeft.hours < 1
```
Comment (kr-matthews) [inline app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:43]:
```suggestion
      { timeLeft.days < 1 && timeLeft.hours < 1
```
otherwise 1 day 0 hours 1 minute 1 second would satisfy this.

File: app/webpacker/components/RegistrationsV2/Register/index.jsx [hunk 4]
```diff
@@ -53,22 +58,39 @@ function Register({
...
+    setTimerEnded(true);
+  }, [setTimerEnded]);
+
+  if (isFetching) {
+    return <Loading />;
+  }
+
+  if (userCanPreRegister(userInfo, competitionInfo) || competitionInfo['registration_currently_open?'] || timerEnded) {
+    return (
+      isFetching ? <Loading />
```
Comment (kr-matthews) [inline app/webpacker/components/RegistrationsV2/Register/index.jsx:71]: The is-fetching case is handled above now, right? Doesn't need to be here.

**Review by kr-matthews — CHANGES_REQUESTED** [review]

File: app/webpacker/components/RegistrationsV2/Register/index.jsx (hunk 4 above)
Comment (FinnIckler) [reply app/webpacker/components/RegistrationsV2/Register/index.jsx:71]: copy paste whoopsie

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx (hunk 2 above)
Comment (FinnIckler) [reply app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:35]: mh.. I don't really expect anyone to sit on the page if it's more than an hour away

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx (hunk 1 above)
Comment (FinnIckler) [reply app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:11]: I call updateTimer immediately in the useEffect. isn't that enough? I haven't seen any rendering issues on local

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx (hunk 2 above)
Comment (FinnIckler) [reply app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:35]: I guess `const [timeLeft, setTimeLeft] = useState(start.diff(DateTime.local(), ['days', 'hours', 'minutes', 'seconds']).toObject());` this would be enough though

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx (hunk 1 above)
Comment (FinnIckler) [reply app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:11]: I guess `const [timeLeft, setTimeLeft] = useState(start.diff(DateTime.local(), ['days', 'hours', 'minutes', 'seconds']).toObject());` this would be enough though

Comment (kr-matthews) [reply app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:11]: It's working now, but if someone comes along and modifies the effect, there's a chance of them introducing a bug. For example, any new effects which depend on these values will be triggered on the initial render when they're all 0.
If something can be done without an effect (such as setting this initial value) then it's almost always better to do without.

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+  // Update the timer every second
+  useEffect(() => {
+    const updateTimer = () => {
+      const now = DateTime.local();
+
+      if (now >= start) {
+        onTimerEnd();
+        return;
+      }
```
Comment (kr-matthews) [inline app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:24]: Not a big deal, but it's best to have each effect do 1 effect, so I would separate this onto its own effect (dependent on all the time values - if they're all 0 then call `onTimerEnd`). (Which is an example of why they should be initialized properly.)

Comment (FinnIckler) [reply app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx]: I added the second effect

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx (hunk 1 above)
Comment (FinnIckler) [reply app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx]: makes sense, I changed it! Will retest on monday

**Comment by FinnIckler** [comment]
still works!

**Comment by dunkOnIT** [comment]
@kr-matthews just checking if your requested changes have been addressed?

**Comment by kr-matthews** [comment]
> @kr-matthews just checking if your requested changes have been addressed?

yes

**Comment by dunkOnIT** [comment]
Awesome! If you mark you changes as resolved/submit a review for the React side of things that would be great

**Review by kr-matthews — APPROVED** [review]
Didn't test locally, but the react code looks reasonable

File: app/models/competition.rb [hunk 6]
```diff
@@ -732,7 +732,7 @@ def uses_new_registration_service?
   end
 
   def should_render_register_v2?(user)
-    uses_new_registration_service? && user.cannot_register_for_competition_reasons(self).empty? && (registration_currently_open? || user_can_pre_register?(user))
```
Comment (gregorbg) [inline app/models/competition.rb:735]: Why is this change necessary, what does it do in the context of the PR?

File: app/views/registrations/register.html.erb [hunk 7]
```diff
@@ -10,7 +10,7 @@
   <% opens_in_days = distance_of_time_in_words_to_now(@competition.registration_open) %>
   <% closes_in_days = distance_of_time_in_words_to_now(@competition.registration_close) %>
 
-  <% if @competition.registration_not_yet_opened? %>
+  <% if !@competition.uses_new_registration_service? && @competition.registration_not_yet_opened? %>
```
Comment (gregorbg) [inline app/views/registrations/register.html.erb:13]: I suspect this is womehow linked to the above change in `should_render_register_v2?`

File: app/webpacker/components/RegistrationsV2/Register/CompetingStep.jsx [hunk 8]
```diff
@@ -31,7 +32,7 @@ const maxCommentLength = 240;
 const potentialWarnings = (competitionInfo) => {
   const warnings = [];
   // Organizer Pre Registration
-  if (!competitionInfo['registration_currently_open?']) {
+  if (DateTime.fromISO(competitionInfo.registration_open) > DateTime.now()) {
```
Comment (gregorbg) [inline app/webpacker/components/RegistrationsV2/Register/CompetingStep.jsx:35]: Too much parsing logic inside an `if` clause. I feel like we already have a method for that, let me check...

Comment (gregorbg) [reply app/webpacker/components/RegistrationsV2/Register/CompetingStep.jsx:35]: Yes, we have `hasPassed` and `hasNotPassed` in `lib/utils/dates.js`

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+import React, { useEffect, useState } from 'react';
+import { Message } from 'semantic-ui-react';
+import I18nHTMLTranslate from '../../I18nHTMLTranslate';
+
+export default function RegistrationClosedMessage({
+  registrationStart,
+  onTimerEnd,
+}) {
+  const start = DateTime.fromISO(registrationStart);
+  const [timeLeft, setTimeLeft] = useState(start.diff(DateTime.local(), ['days', 'hours', 'minutes', 'seconds']).toObject());
```
Comment (gregorbg) [inline app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:11]: This should definitely be a `util` function

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+  useEffect(() => {
+    if (timeLeft.days === 0
+      && timeLeft.hours === 0
+      && timeLeft.minutes === 0
+      && timeLeft.seconds === 0) {
+      onTimerEnd();
+    }
+  }, [onTimerEnd, timeLeft.days, timeLeft.hours, timeLeft.minutes, timeLeft.seconds]);
```
Comment (gregorbg) [inline app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:20]: Can you just use `timeLeft` as a whole as dependency?

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+  useEffect(() => {
+    const updateTimer = () => {
```
Comment (gregorbg) [inline app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:23]: Why are you defining this as a local function within `useEffect`?

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+      const now = DateTime.local();
+
+      const diff = start.diff(now, ['days', 'hours', 'minutes', 'seconds']).toObject();
```
Comment (gregorbg) [inline app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:26]: Use a shared function, see above

Comment (gregorbg) [reply app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:26]: And on that note, you could apply `Math.floor` inside that util function already, so that you can use the return value of the function as param to `setTimeLeft` directly below

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    updateTimer(); // Initial call to set the state right away
+
+    return () => clearInterval(intervalId); // Cleanup interval on unmount
+  }, [registrationStart, onTimerEnd, start]);
+
+  return (
+    <Message color="blue">
+      { timeLeft.days < 1 && timeLeft.hours < 1
+        ? <I18nHTMLTranslate i18nKey="competitions.registration_v2.register.will_open_countdown" options={{ minutes: timeLeft.minutes, seconds: timeLeft.seconds }} />
+        : <I18nHTMLTranslate i18nKey="registrations.will_open_html" options={{ days: start.toRelative(), time: DateTime.fromISO(registrationStart).toLocaleString(DateTime.DATETIME_FULL) }} />}
```
Comment (gregorbg) [inline app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:45]: Isn't `const start` above exactly the same as `DateTime.fromISO(registrationStart)` so that you could rewrite this to `start.toLocaleString(...)`?

File: app/webpacker/components/RegistrationsV2/Register/index.jsx [hunk 14]
```diff
@@ -8,6 +8,9 @@ import StoreProvider, { useDispatch } from '../../../lib/providers/StoreProvider
 import messageReducer from '../reducers/messageReducer';
 import WCAQueryClientProvider from '../../../lib/providers/WCAQueryClientProvider';
 import ConfirmProvider from '../../../lib/providers/ConfirmProvider';
+import RegistrationClosedMessage from './RegistrationClosedMessage';
+
+const userCanPreRegister = (user, competitionInfo) => competitionInfo.organizers.some((o) => o.id === user.id);
```
Comment (gregorbg) [inline app/webpacker/components/RegistrationsV2/Register/index.jsx:13]: Why only organizers and not Delegates?

File: app/webpacker/components/RegistrationsV2/Register/index.jsx [hunk 15]
```diff
@@ -53,22 +58,36 @@ function Register({
...
 
+  const onTimerEnd = useCallback(() => {
+    setTimerEnded(true);
+  }, [setTimerEnded]);
+
+  if (isFetching) {
+    return <Loading />;
+  }
+
+  if (userCanPreRegister(userInfo, competitionInfo) || competitionInfo['registration_currently_open?'] || timerEnded) {
```
Comment (gregorbg) [inline app/webpacker/components/RegistrationsV2/Register/index.jsx:69]: Lots of calculations inside one `if` statement, not happy about that

**Review by gregorbg — CHANGES_REQUESTED** [review]
I feel like you're doing three changes in this PR.
1. Reduce the amount of data returned by the internal `User` API
2. Implement a timer for registration opening
3. Enable pre-registration for competition managers

(1) Is very trivial but important, and it's unfortunate that this change is being blocked by discussions about (2) and (3). So it should be its own PR.

(3) needs its own design discussion, especially because I think it depends on another PR in the microservice. If that PR gets delayed, it would also block features (1) and (2), so it should be its own PR.

(2) is the core essence of this PR, and I have added lots of comments. Please split out the other things into separate PRs

File: app/models/competition.rb (hunk 6 above)
Comment (FinnIckler) [reply app/models/competition.rb:735]: Because before we showed the Registration Opens in x in Ruby only and now we want to show it in react

File: app/views/registrations/register.html.erb (hunk 7 above)
Comment (FinnIckler) [reply app/views/registrations/register.html.erb:13]: yes

File: app/webpacker/components/RegistrationsV2/Register/CompetingStep.jsx (hunk 8 above)
Comment (FinnIckler) [reply app/webpacker/components/RegistrationsV2/Register/CompetingStep.jsx:35]: done

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx (hunk 10 above)
Comment (FinnIckler) [reply app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:20]: I had it auto generated by Rubymine, but it looks like eslint doesn't complain if I just use the whole object (also I googled and it looks like it's only an issue if the reference changes, with I think isn't the case here?)

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx (hunk 12 above)
Comment (FinnIckler) [reply app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:26]: done

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx (hunk 13 above)
Comment (FinnIckler) [reply app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:45]: true! must have been before I refactored it

File: app/webpacker/components/RegistrationsV2/Register/index.jsx (hunk 15 above)
Comment (FinnIckler) [reply app/webpacker/components/RegistrationsV2/Register/index.jsx:69]: should be better when I move userCanPreRegister to Ruby?

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+import { Message } from 'semantic-ui-react';
+import I18nHTMLTranslate from '../../I18nHTMLTranslate';
+import { fullTimeDiff } from '../../../lib/utils/dates';
+
+export default function RegistrationClosedMessage({
+  registrationStart,
+  onTimerEnd,
+}) {
+  const start = DateTime.fromISO(registrationStart);
+  const [timeLeft, setTimeLeft] = useState(start.diff(DateTime.local(), ['days', 'hours', 'minutes', 'seconds']).toObject());
```
Comment (gregorbg) [inline app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx:12]: Why is this not using the new `fullTimeDiff` util function, now that we have it?

**Review by gregorbg — CHANGES_REQUESTED** [review]

File: app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx (hunk 16 above)
Comment (FinnIckler) [reply app/webpacker/components/RegistrationsV2/Register/RegistrationClosedMessage.jsx]: good catch

**Review by gregorbg — APPROVED** [review]

**Comment by FinnIckler** [comment]
I'll do one more test and then I'll merge it

**Comment by FinnIckler** [comment]
confirmed still working
